mautic/mautic
# Add additional checks to installation + update process

~⚠️ **This PR needs #10481 to be merged first** ⚠️~ Done ✅ 

| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | 4.x
| Bug fix?                               | no
| New feature?                           | yes
| Deprecations?                          | no
| BC breaks?                             | no
| Automated tests included?              | yes
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Related to #9607, fixes #10472




#### Description:

As described in https://github.com/mautic/mautic/issues/9607, we wanted to have additional checks in the Mautic installation and update process, like PHP version, database driver/version, etc.

This PR lays the foundation for pre-update checks. Those checks are executed both in the UI and CLI update processes. This looks as follows (ignore the actual versions shown in the screenshots):

**Pre-update checks in the UI**

![image](https://user-images.githubusercontent.com/17739158/134695440-db341cca-3585-4e6e-aa2c-52d808c85133.png)

**Pre-update checks in the CLI**

![image](https://user-images.githubusercontent.com/17739158/134695545-451b009e-18be-4210-867d-e80a04520255.png)

New pre-update checks can be added easily by creating a new check in `app/bundles/CoreBundle/Helper/Update/PreUpdateChecks`, then tagging it with `mautic.update_check` in the bundle's `config.php`. Then it will be picked up as a pre-update check.

This PR adds the following checks as a starting point:
- Database driver (MariaDB or MySQL)
- Database version (MySQL 5.7.14+ or MariaDB 10.2.7+) --> these can now be configured dynamically in `app/release_metadata.json`)
- PHP version (`minimum_php_version` and `maximum_php_version` in `app/release_metadata.json`)

**Other fixes/changes in this PR**
- Fixes #10472 where the user wasn't informed at all if a newer Mautic version was available, but the user's PHP version was too low. The user now sees that a new Mautic version is available, but the update will fail at the pre-update check step, telling the user that their PHP version is too old.
- Don't run `upload-release-asset` GitHub action on forks (created a test release for this PR at https://github.com/dennisameling/mautic/releases/tag/4.9.99-alpha so ran into this issue there)
- Remove a bunch of legacy code in `upgrade.php` that's no longer used
- Fixed bug where the CLI installation defaults to pdo_mysql (`db_driver`) if a database driver is set in `app/config/local.php` but not in the CLI options. The same goes for `db_backup_tables` and `db_backup_prefix`.
- Now leverages more [PHP typed properties](https://stitcher.io/blog/typed-properties-in-php-74) (this is possible since we require PHP 7.4 as of Mautic 4)



#### Steps to test this PR:
1. Checkout this PR locally: `gh pr checkout 10459`
2. For the remainder of the steps, we assume that you're using DDEV for local development. This allows you to easily change your PHP/database versions 
3. **Before testing each scenario below**, please ensure you have a clean state where Mautic is not installed: `ddev delete --omit-snapshot --yes && rm -rf var/cache && rm app/config/local.php`
4. Before every test run you do, set the correct PHP/database versions in the DDEV config:
  - Open `.ddev/config.yaml`
  - Change `php_version` (don't forget to update the version in `webimage_extra_packages` as well), `mariadb_version`, `mysql_version` as needed. NOTE: you can only have EITHER a `mariadb_version` OR a `mysql_version` active, not both!
5. Run `ddev restart` so that the appropriate PHP/database versions will be activated.

For testing the **CLI installation**, please run `ddev exec php bin/console mautic:install https://mautic.ddev.site --db_driver=pdo_mysql --db_host=db --db_user=db --db_name=db --db_password=db --db_port=3306 --admin_password=mautic --admin_email="demo@ddev.local"` - you can change the database driver by changing `db_driver` accordingly in this command.

**Here's the text matrix for the installation process - all scenarios below have to be tested.** 

| Action | PHP version | Database driver | Database version | Expected outcome
|--|--|--|--|--|
| Install using UI | 7.3.x | pdo_mysql | 10.3.x | ❌ on PHP version
| Install using CLI | 7.3.x | pdo_mysql | 10.3.x | ❌ on PHP version
| Install using CLI | 7.4.x | pdo_sqlite | 10.3.x | ❌ on database driver
| Install using UI | 7.4.x | pdo_mysql | MariaDB 10.1.x | ❌ on database version
| Install using CLI | 7.4.x | pdo_mysql | MariaDB 10.1.x | ❌ on database version
| Install using UI | 7.4.x | pdo_mysql | MySQL 5.6.x | ❌ on database version
| Install using CLI | 7.4.x | pdo_mysql | MySQL 5.6.x | ❌ on database version
| Install using UI | 7.4.x | pdo_mysql | MariaDB 10.3.x | ✅ 
| Install using CLI | 7.4.x | pdo_mysql | MariaDB 10.3.x | ✅ 
| Install using UI | 7.4.x | pdo_mysql | MySQL 5.7.x | ✅ 
| Install using CLI | 7.4.x | pdo_mysql | MySQL 5.7.x | ✅ 

Once you have a working Mautic installation, we can start testing the update scenario as well. **Before you start testing the update scenario, please head over to `app/config/local.php` and add the following lines at the bottom**:

```
'system_update_url' => 'https://api.github.com/repos/dennisameling/mautic/releases',
'update_stability' => 'alpha',
```

Then run `rm -rf var/cache` to remove the cache and `ddev exec "php bin/console mautic:update:find"` to check for updates. Please only proceed if you see "Version 4.9.99-alpha of Mautic is available for download". I bumped the minimum PHP version to 8.0.0, MySQL version to 8.0.0 and MariaDB version to 10.5.0 so that you can see the update errors show up.

- You can try the **CLI** update by running `ddev exec "php bin/console mautic:update:apply"`.
- You can try the **UI** update by going to https://mautic.ddev.site/index_dev.php/s/update in your browser (it's important you use the dev version as the production assets will only be generated when we release a new Mautic version)

| Action | PHP version | Database driver | Database version | Expected outcome
|--|--|--|--|--|
| Update using UI | 7.4.x | pdo_mysql | 10.3.x | ❌ on PHP version and database version
| Update using CLI | 7.4.x | pdo_mysql | 10.3.x | ❌ on PHP version and database version




<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10459"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
